pylint/extensions/redefined_loop_name.py

```diff
@@ -0,0 +1,87 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Optional checker to warn when loop variables are overwritten in the loop's body."""
+
+from __future__ import annotations
+
+from astroid import nodes
+
+from pylint import checkers
+from pylint.checkers import utils
+from pylint.interfaces import HIGH
+from pylint.lint import PyLinter
+
+
+class RedefinedLoopName(checkers.BaseChecker):
+
+    name = "redefined-loop-name"
+
+    msgs = {
+        "W2901": (
+            "Redefining %r from loop (line %s)",
+            "redefined-loop-name",
+            "Used when a loop variable is overwritten in the loop body.",
+        ),
+    }
+
+    def __init__(self, linter: PyLinter) -> None:
+        super().__init__(linter)
+        self._loop_variables: list[tuple[nodes.For, list[str | None]]] = []
```

Comment:
```suggestion
        self._loop_variables: list[tuple[nodes.For, list[str]]] = []
```

I think. `mypy` will complain because `.name` is typed as `str | None` I think, but that is a mistake. An `Assign` node should always have a `str` `name` attribute. Generally, whenever typing in `astroid` is incorrect we add `type: ignore` to offending lines instead of "trickling-down" the incorrect typing to `pylint`.

_I have also stopped using assertions in these cases as it is difficult to keep track of those and remove them when we update `astroid`. We get warned about unused ignores, so that's often the better option imo._